### PR TITLE
[Enhancement] Fix zero points exercises with and without bonus for all exercise types (manual assessment + complaints)

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested all changes and all related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x]  Server: I followed the coding and design guidelines.
- [x]  Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x]  Server: I implemented the changes with a good performance and prevented too many database calls
- [x]  Server: I documented the Java code using JavaDoc style.
- [x]  Client: I followed the coding and design guidelines.
- [x]  Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x]  Client: I documented the TypeScript code using JSDoc style.
- [x]  Client: I added multiple screenshots/screencasts of my UI changes
- [x]  Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Currently zero point exercises are not completely supported on all exercise types for manual assessment and complaints. This PR introduces the support for this.

### Description

- [x] Extract `getMaxScoreRespectingZeroPointExercises` so that it can be used from other components on server and client
- [x] Implement correct use of maxPoints for assessment of Modeling, Text and FileUpload exercises + complaints
- [x] Implement correct use of maxPoints for assessment of Programming exercises + complaints

### Steps for Testing

Normal exercises:

1. Log in to Artemis
2. Navigate to Course Administration

**Important: test the following 4 cases in 4 different exercises for Programming, Modeling, Text, Quiz and FileUpload:**
 - [x] 10 points + 0 bonus points
 - [x] 10 points + 10 bonus points
 - [x] 0 points + 10 bonus points
 - [x] 0 points + 0 bonus points

Test the following scenario: if a student e.g. has collected 30 points in a course, but the sum of normal points is 20, then the score on the start page of Artemis should show 150%. The score calculation is really critical and has to be absolutely correct!
Also make sure that the instructor page course scores includes the correct calculation based on the sum of normal points and the exported csv file is correct.

3. Assess the exercises 
4. See if everything works as expected
5. Complain and assess the complaint
6. See if everything works as expected

### Test Coverage

Server:

- Exercise.java: 92%
- Result.java: 95%
- AssessmentService.java: 88%
- ProgrammingAssessmentService.java: 92%
- ProgrammingExerciseGradingService.java: 95%

Client:

- scoreChartPreset.ts: 92%
- exercise-utils.ts: 91%
- result-detail.component.ts: 95%
- code-editor-tutor-assessment-container.component: 81%
- text-feedback-conflicts.component.ts: 79%
- modeling-assessment.component.ts: 78.95%
- score-display.component: 76%
- modeling-assessment-editor.component.ts: 72%
- text-assessment-base.component.ts: 71%
- text-submission-assessment.component: 70.11%
- modeling-submission.component.ts:  72.5%
- file-upload-assessment.component.ts: 73.44%

### Screenshots

Tooltip when bonus points are available:
image

Statistics view of an exercise:
image

0 point and 10 bonus point exercise:
image

0 point and 0 bonus point exercise (showing only the percentage reached, base is 100 points):
image

### CODE REVIEW

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
Code for this PR looks good, but when working on assessing the text/ modelling exercises with bonus points there occurs a bug: 
Those exercises only have bonus points and no normal points. We cannot add bonus points during the assessment, therefore we gave normal points (different amounts < maximum bonus points). The assessment dashboard and the students view then shows a totally wrong percentage for the grading. This happened with different amounts of (bonus) points given and always evaluated to the same percentage value.
Screenshot 2020-12-19 at 12 37 00
Programming exercises did not have this problem in our testing session with @simonzilker , @airelawaleria and @marvinchimeka .

**Review by simonzilker — CHANGES_REQUESTED** [review]
What @JohannesWeiss said.
Scoreboard for Modeling exercise with 0 Points and 10 Bonus Points: 
image

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested in the testing session. Confirm the issue above.
image

**Review by marvinchimeka — COMMENTED** [review]
tested in a testing session. The case where you have 0 points and 10 Bonus points seems to work only for programming exercises. Agreeing with the mentioned issue by @JohannesWeiss .

**Review by EneaGore — CHANGES_REQUESTED** [review]
Can confirm same issue.
IssueBonus

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 1]
```diff
@@ -81,10 +81,10 @@ Result submitResult(Result result, Exercise exercise, Double calculatedScore) {
 
         result.setCompletionDate(ZonedDateTime.now());
         // Take bonus points into account to achieve a result score > 100%
-        double totalScore = calculateTotalScore(calculatedScore, maxScore + bonusPoints);
+        double totalScore = calculateTotalScore(calculatedScore, maxPoints);
         // Set score and resultString according to maxScore, to establish results with score > 100%
-        result.setScore(totalScore, maxScore);
-        result.setResultString(totalScore, maxScore);
+        result.setScore(totalScore, maxPoints);
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:86]:
```suggestion
        result.setScore(totalScore, maxPointsRespectingZeroPoints);
```

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 2]
```diff
@@ -81,10 +81,10 @@ Result submitResult(Result result, Exercise exercise, Double calculatedScore) {
...
+        result.setResultString(totalScore, maxPoints);
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:87]:
```suggestion
        result.setResultString(totalScore, maxPointsRespectingZeroPoints);
```

**Comment by derLalla** [comment]
> What @JohannesWeiss said.
> Scoreboard for Modeling exercise with 0 Points and 10 Bonus Points:
> image

this is fixed with the latest changes

**Review by simonzilker — CHANGES_REQUESTED** [review]
Tested in testing session with @airelawaleria @JohannesWeiss @LudwigWeide . Exercises with 0 normal Points and 0 Bonus Points don´t work for all exercise types as you can see in the picture
image

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested in the testing session. 0 points 0 bonus points test case fails for every exercise type.

**Review by LudwigWeide — CHANGES_REQUESTED** [review]
Tested with @simonzilker . Same issue.

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
Tested again on ts1 in testing session and con confirm the issue mentioned by @simonzilker.
Additionally, we found a discrepancy in the percentage values in the course overview and the statistics tab of the course (116% and 133% in our case) that only gets updated, if one presses "Refresh". 

output

**Review by marvinchimeka — CHANGES_REQUESTED** [review]
The Test case with 0 points 0 bonus does not work for all exercise types .
Confirming already mentioned issue.

**Comment by derLalla** [comment]
> Tested in testing session with @airelawaleria @JohannesWeiss @LudwigWeide . Exercises with 0 normal Points and 0 Bonus Points don´t work for all exercise types as you can see in the picture
> image

what is the issue? Do you mean that you have " x points of 100 points"?  
This is intended, as 0 point/0 bonus point exercises have a default of 100 points. So a user can achieve up to 100% of a the score.

**Comment by stefanwaldhauser** [comment]
> Tested in testing session with @airelawaleria @JohannesWeiss @LudwigWeide . Exercises with 0 normal Points and 0 Bonus Points don´t work for all exercise types as you can see in the picture
> image

I am not sure what do you mean? 

image
image (1)

**Review by airelawaleria — APPROVED** [review]
Since the above mentioned behavior is normal, re-approve

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve as well.

**Comment by stefanwaldhauser** [comment]
> Tested again on ts1 in testing session and con confirm the issue mentioned by @simonzilker.
> Additionally, we found a discrepancy in the percentage values in the course overview and the statistics tab of the course (116% and 133% in our case) that only gets updated, if one presses "Refresh".
> 
> output

I think I fixed this. Can you check again please :) ?

**Comment by stefanwaldhauser** [comment]
> > Tested again on ts1 in testing session and con confirm the issue mentioned by @simonzilker.
> > Additionally, we found a discrepancy in the percentage values in the course overview and the statistics tab of the course (116% and 133% in our case) that only gets updated, if one presses "Refresh".
> > output
> 
> I think I fixed this. Can you check again please :) ?

Reverted fix and will be dealt in a follow up.

**Review by airelawaleria — APPROVED** [review]
Tested in the testing session @marvinchimeka @JohannesWeiss @simonzilker @LudwigWeide . Works fine

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1 in testing session, works as expected so far.

**Review by simonzilker — APPROVED** [review]
Tested in testing session. Found no bugs! Looks good!

**Review by marvinchimeka — APPROVED** [review]
tested in testing session. Works as intended.

**Review by EneaGore — APPROVED** [review]
Seems to be working fine on ts1 now.

**Review by valeriebucher — APPROVED** [review]
Looks fine to me. Worked as described on ts1

**Review by tiffca — APPROVED** [review]
Tested successfully on ts1. 

noticed one behavior on file upload where the points says out of 100 points for some reason (even though it's out of 0 points and calculated correctly), but not a big issue.
Reproduce steps: 
1. Create a file upload exercise with 0 points
2. assess and give 0 points
3. in the student view look at how many points they have. Actual: it says 0 out of 100. Expected: it should say just 0 points or 0 out of 10 or 20 (if bonus included). 

image

**Review by krusche — CHANGES_REQUESTED** [review]
I am also confused by the result string in case of 0 points exercises with 0 bonus points.

image

We should not show anything about points in the result string here.
For programming exercises, please only show the percentage based on the tests
For modeling|text|fileupload exercises, please show a different type of result string, I would suggest some kind of summary about the feedback elements, the minimal solution would be "X feedback"

**Comment by derLalla** [comment]
> I am also confused by the result string in case of 0 points exercises with 0 bonus points.
> 
> image
> 
> We should not show anything about points in the result string here.
> For programming exercises, please only show the percentage based on the tests
> For modeling|text|fileupload exercises, please show a different type of result string, I would suggest some kind of summary about the feedback elements, the minimal solution would be "X feedback"

We have addressed the issue and we are displaying now: 

x%, 0 of 0 points.

**Comment by derLalla** [comment]
> Tested again on ts1 in testing session and con confirm the issue mentioned by @simonzilker.
> Additionally, we found a discrepancy in the percentage values in the course overview and the statistics tab of the course (116% and 133% in our case) that only gets updated, if one presses "Refresh".
> 
> output

this will be addressed in a follow-up PR. Where we want to use websockets to update the view.

**Review by tiffca — APPROVED** [review]
Tested the previously mentioned scenario and it works as expected now, hence approving!

**Review by airelawaleria — APPROVED** [review]
Re-tested on ts3. Re-approve

**Review by simonzilker — DISMISSED** [review]
Looks good! Re-approve.

**Review by JohannesWeiss — DISMISSED** [review]
Looks good to me, code as well. Re-Approve!

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 3]
```diff
@@ -822,4 +822,42 @@ public String getMappedColumnName() {
...
+     * @return true or false
+     */
+    @JsonIgnore
+    public boolean isZeroPointExercise() {
+        boolean hasNormalPoints = Objects.requireNonNullElse(getMaxScore(), 0.0) > 0.0;
+        boolean hasBonusPoints = Objects.requireNonNullElse(getBonusPoints(), 0.0) > 0.0;
+        if (!hasNormalPoints && !hasBonusPoints) {
+            return true;
+        }
+        return false;
```
Comment (philippbauch) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:860]:
```suggestion
        return !hasNormalPoints && !hasBonusPoints;
```

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 4]
```diff
@@ -127,22 +127,38 @@ public void setResultString(String resultString) {
...
         if (maxScore == null) {
             return formatter.format(totalScore) + " points";
         }
         else {
+            if (exercise != null && exercise.isZeroPointExercise()) {
+                totalScore = 0.0;
+                maxScore = 0.0;
+            }
             return formatter.format(totalScore) + " of " + formatter.format(maxScore) + " points";
         }
```
Comment (philippbauch) [inline src/main/java/de/tum/in/www1/artemis/domain/Result.java:163]:
```suggestion
        if (exercise != null && exercise.isZeroPointExercise()) {
            totalScore = 0.0;
            maxScore = 0.0;
        }
        return formatter.format(totalScore) + " of " + formatter.format(maxScore) + " points";
```

You probably don't need that `else`

**Review by philippbauch — CHANGES_REQUESTED** [review]
Added two small suggestions. Otherwise, the code looks good to me!

**Review by JannisBittlmayer — CHANGES_REQUESTED** [review]
Tested on TS1, works apart from some small issues:
- For exercises with 0 normal and 0 bonus points: When you assess this exercise with >0 points (for example 14), the score is set to 14%, 0 of 0 points. This may be intended but looks weird to me
- For File Upload exercises: When assessing this exercise, you cannot give 0 points for "Add feedback" since Artemis gives the warning that "The score field must be a number and can not be empty".
- For exercises with 0 points and (for example) 10 bonus points: When you give 9 points the result is 90%, 9 of 10 points. Shouldn't it be 9 of 0 points? When you have an exercise with 10 normal and 10 bonus points, it can show 15 of 10 points, too. I can see the problem of calculating a percentage for e.g. 9/0 though. I hope it's clear what I mean.

**Review by FlorianStang — CHANGES_REQUESTED** [review]
Tested in TS1 with @JannisBittlmayer, i can confirm the points mentioned by Jannis.

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java (hunk 4 above)
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/domain/Result.java:163]: 👍

**Comment by derLalla** [comment]
> Tested on TS1, works apart from some small issues:
> 
> * For exercises with 0 normal and 0 bonus points: When you assess this exercise with >0 points (for example 14), the score is set to 14%, 0 of 0 points. This may be intended but looks weird to me
> * For File Upload exercises: When assessing this exercise, you cannot give 0 points for "Add feedback" since Artemis gives the warning that "The score field must be a number and can not be empty".
> * For exercises with 0 points and (for example) 10 bonus points: When you give 9 points the result is 90%, 9 of 10 points. Shouldn't it be 9 of 0 points? When you have an exercise with 10 normal and 10 bonus points, it can show 15 of 10 points, too. I can see the problem of calculating a percentage for e.g. 9/0 though. I hope it's clear what I mean.

I see your concerns, but all of that is intended:  
1) for 0 point / 0 bonus point exercises we still need a way to grad and give a score. So 100 is the base for those exercises. This means the shown points are actually percentage points.  

2) You cannot give 0 points for a Feedback, this wouldn't make sense. I guess you are confused by the meaning of a zero point exercise. As mentioned in 1) this should be seen as a percentage score.  

3) We always want to show the maximum points. So for a 10/10 exercise we are showing 20 of 10 points (Bonus: 10).  
For a 0/10 exercise, it is 10 of 10 points (Bonus: 10)
For a 10/0 exercise, it is 10 of 10 points.
For a 0/0 exercise, it is 0 of 0 points.

**Comment by derLalla** [comment]
> Tested in TS1 with @JannisBittlmayer, i can confirm the points mentioned by Jannis.

see my explanation above :)

**Comment by derLalla** [comment]
> Added two small suggestions. Otherwise, the code looks good to me!

thanks, I have committed your changes 👍

**Comment by JannisBittlmayer** [comment]
> 
> 
> > Tested on TS1, works apart from some small issues:
> > 
> > * For exercises with 0 normal and 0 bonus points: When you assess this exercise with >0 points (for example 14), the score is set to 14%, 0 of 0 points. This may be intended but looks weird to me
> > * For File Upload exercises: When assessing this exercise, you cannot give 0 points for "Add feedback" since Artemis gives the warning that "The score field must be a number and can not be empty".
> > * For exercises with 0 points and (for example) 10 bonus points: When you give 9 points the result is 90%, 9 of 10 points. Shouldn't it be 9 of 0 points? When you have an exercise with 10 normal and 10 bonus points, it can show 15 of 10 points, too. I can see the problem of calculating a percentage for e.g. 9/0 though. I hope it's clear what I mean.
> 
> I see your concerns, but all of that is intended:
> 
>     1. for 0 point / 0 bonus point exercises we still need a way to grad and give a score. So 100 is the base for those exercises. This means the shown points are actually percentage points.
> 
>     2. You cannot give 0 points for a Feedback, this wouldn't make sense. I guess you are confused by the meaning of a zero point exercise. As mentioned in 1) this should be seen as a percentage score.
> 
>     3. We always want to show the maximum points. So for a 10/10 exercise we are showing 20 of 10 points (Bonus: 10).
>        For a 0/10 exercise, it is 10 of 10 points (Bonus: 10)
>        For a 10/0 exercise, it is 10 of 10 points.
>        For a 0/0 exercise, it is 0 of 0 points.

Okay, thanks for the quick response :)

**Comment by derLalla** [comment]
> > > Tested on TS1, works apart from some small issues:
> > > 
> > > * For exercises with 0 normal and 0 bonus points: When you assess this exercise with >0 points (for example 14), the score is set to 14%, 0 of 0 points. This may be intended but looks weird to me
> > > * For File Upload exercises: When assessing this exercise, you cannot give 0 points for "Add feedback" since Artemis gives the warning that "The score field must be a number and can not be empty".
> > > * For exercises with 0 points and (for example) 10 bonus points: When you give 9 points the result is 90%, 9 of 10 points. Shouldn't it be 9 of 0 points? When you have an exercise with 10 normal and 10 bonus points, it can show 15 of 10 points, too. I can see the problem of calculating a percentage for e.g. 9/0 though. I hope it's clear what I mean.
> > 
> > 
> > I see your concerns, but all of that is intended:
> > ```
> > 1. for 0 point / 0 bonus point exercises we still need a way to grad and give a score. So 100 is the base for those exercises. This means the shown points are actually percentage points.
> > 
> > 2. You cannot give 0 points for a Feedback, this wouldn't make sense. I guess you are confused by the meaning of a zero point exercise. As mentioned in 1) this should be seen as a percentage score.
> > 
> > 3. We always want to show the maximum points. So for a 10/10 exercise we are showing 20 of 10 points (Bonus: 10).
> >    For a 0/10 exercise, it is 10 of 10 points (Bonus: 10)
> >    For a 10/0 exercise, it is 10 of 10 points.
> >    For a 0/0 exercise, it is 0 of 0 points.
> > ```
> 
> Okay, thanks for the quick response :)

btw. Do not hit "Close with comment", as this will close the whole PR :D

**Comment by JannisBittlmayer** [comment]
> 
> 
> > > > Tested on TS1, works apart from some small issues:
> > > > 
> > > > * For exercises with 0 normal and 0 bonus points: When you assess this exercise with >0 points (for example 14), the score is set to 14%, 0 of 0 points. This may be intended but looks weird to me
> > > > * For File Upload exercises: When assessing this exercise, you cannot give 0 points for "Add feedback" since Artemis gives the warning that "The score field must be a number and can not be empty".
> > > > * For exercises with 0 points and (for example) 10 bonus points: When you give 9 points the result is 90%, 9 of 10 points. Shouldn't it be 9 of 0 points? When you have an exercise with 10 normal and 10 bonus points, it can show 15 of 10 points, too. I can see the problem of calculating a percentage for e.g. 9/0 though. I hope it's clear what I mean.
> > > 
> > > 
> > > I see your concerns, but all of that is intended:
> > > ```
> > > 1. for 0 point / 0 bonus point exercises we still need a way to grad and give a score. So 100 is the base for those exercises. This means the shown points are actually percentage points.
> > > 
> > > 2. You cannot give 0 points for a Feedback, this wouldn't make sense. I guess you are confused by the meaning of a zero point exercise. As mentioned in 1) this should be seen as a percentage score.
> > > 
> > > 3. We always want to show the maximum points. So for a 10/10 exercise we are showing 20 of 10 points (Bonus: 10).
> > >    For a 0/10 exercise, it is 10 of 10 points (Bonus: 10)
> > >    For a 10/0 exercise, it is 10 of 10 points.
> > >    For a 0/0 exercise, it is 0 of 0 points.
> > > ```
> > 
> > 
> > Okay, thanks for the quick response :)
> 
> btw. Do not hit "Close with comment", as this will close the whole PR :D

I realized this after posting, sorry! Won't happen again!

**Review by JannisBittlmayer — APPROVED** [review]
Everything works as intended.

**Review by JohannesWeiss — APPROVED** [review]
New code changes lgtm, therefore re-approve!

**Review by airelawaleria — APPROVED** [review]
Re-approve

**Review by simonzilker — APPROVED** [review]
Re-approve

**Review by marvinchimeka — DISMISSED** [review]
Looks Good. Re-approve

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts [hunk 5]
```diff
@@ -110,6 +110,7 @@ export class StudentExamDetailComponent implements OnInit {
                 exercise.studentParticipations[0].results?.length &&
                 exercise.studentParticipations[0].results.length > 0
             ) {
+                // TODO: rene: is it possible to have only bonusPoint exams?
```
Comment (FuchsDominik) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:113]: Was this forgotten or left on purpose?

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 6]
```diff
@@ -502,6 +503,7 @@ export class CodeEditorTutorAssessmentContainerComponent implements OnInit, OnDe
         this.manualResult!.hasFeedback = true;
         // Append the automatic result string which the manual result holds with the score part, to create the manual result string
         if (this.isFirstAssessment) {
+            // TODO: rene: change the resultString depending on bonusPoints?
```
Comment (FuchsDominik) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:506]: Is this relevant for follow-ups?

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 7]
```diff
@@ -510,11 +512,12 @@ export class CodeEditorTutorAssessmentContainerComponent implements OnInit, OnDe
              */
             const resultStringParts: string[] = this.manualResult!.resultString!.split(', ');
             // When no maxScore is set, then show only the achieved points
+            // TODO: rene: change the resultString depending on bonusPoints?
```
Comment (FuchsDominik) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:515]: Same here?

File: src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.ts [hunk 8]
```diff
@@ -334,7 +335,11 @@ export class ExampleTextSubmissionComponent implements OnInit, AfterViewInit {
             return;
         }
 
-        this.totalScore = credits.reduce((a, b) => a! + b!, 0)!;
+        // this.totalScore = credits.reduce((a, b) => a! + b!, 0)!;
```
Comment (FuchsDominik) [inline src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.ts:338]: Is not needed i assume?

**Review by FuchsDominik — APPROVED** [review]
Code looks good to me! Really just found some questionable comments which i came across

**Review by tiffca — APPROVED** [review]
Everything worked for me, except I couldn't create programming exercises - probably some weird bug I'm experiencing, hence approving.
The other exercises worked fine.

**Review by airelawaleria — APPROVED** [review]
Changes look good. Re-approve

**Review by simonzilker — DISMISSED** [review]
looks good!

**Review by marvinchimeka — APPROVED** [review]
Works fine and code looks good.

**Review by JohannesWeiss — DISMISSED** [review]
Looks good to me, re-approve!

**Review by EneaGore — DISMISSED** [review]
Looks good, for programming exercises it still shows points out of 100 for 0+0 points if you open the Feedback. It is not caluclated into the actual points though.

**Review by nicolasruscher — DISMISSED** [review]
Looking good!

**Review by airelawaleria — DISMISSED** [review]
Tested with @simonzilker on ts1 again. Everything works. Now for 0-0 exercises the precentual achievement is shown together with "0 of 0 points" string.

**Review by simonzilker — DISMISSED** [review]
Tested with @airelawaleria on ts1. Works as intended now!

**Review by LudwigWeide — APPROVED** [review]
Tested on TS with @JohannesWeiss . Works as intended. Point display a little confusing (58% while 0 out of 0 points.) as in Screenshot.
0_0points

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1 with @LudwigWeide. Basically, everything works as intended.
With a 0 0 programming exercise a percentage is shown depending on the given points, but we assume this is not a real issue, since for those exercises no points should be given. Since zero out of zero points are displayed correctly this can still be approved.
image

**Review by marvinchimeka — APPROVED** [review]
Could not find any issue. 0 of 0 string is shown as intended.

**Review by airelawaleria — APPROVED** [review]
Re-approve

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve!

**Review by simonzilker — APPROVED** [review]
re-approve

**Review by LudwigWeide — DISMISSED** [review]
Re-Approve.

**Review by tiffca — APPROVED** [review]
Tested all scenarios with programming exercise and it works!

**Comment by derLalla** [comment]
> Tested on ts1 with @LudwigWeide. Basically, everything works as intended.
> With a 0 0 programming exercise a percentage is shown depending on the given points, but we assume this is not a real issue, since for those exercises no points should be given. Since zero out of zero points are displayed correctly this can still be approved.
> image

This is intended, and going to be addressed in a follow-up.

**Review by airelawaleria — APPROVED** [review]
Re-approve

**Review by simonzilker — APPROVED** [review]
re-approve

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve!

**Review by marvinchimeka — APPROVED** [review]
re-approve

**Review by EneaGore — DISMISSED** [review]
Last working release seems to work as expected.

**Comment by kloessst** [comment]
**Programming Exercise Manual Assessment:**

- With points and bonus the result string only shows the points initially
image

- Without points but with bonus the same max 10 points are shown
noPB

Given this I would expect, that the exercise with points and bonus shows the points + bonus as the max points a tutor can give.

Tutors should be able to see, how many points and bonus points they can give for each programming exercise. This could also be realized as a tooltip.

Students should also be able to examine, how many bonus points an exercise can give here:
image

**Programming Exercise Complaints:**

After the manual assessment, the dashboard shows a trigger build plan button for the exercise without points but with bonus. The automatic assessment of this submission was successful before.
image

Seems like complaints are deactivated for exercises without points and bonus. This makes sense but we should explain this in the tooltip:
image

Need to verify this findings as TS1 got redeployed at some point.

**Review by FrankeLukas — COMMENTED** [review]
> See if everything works as expected

What is the expected Score for 0 point exercises? I created a new text exercise with 0 points and 0 bonus points, then assessed a submission with 1 point which results in the following:
grafik
I assume that is correct because we use a fallback of `100` as base when there are no points. However, on the course statistics for the student I still have 0 points? (When in fact I should have 1 point?)
grafik

**Comment by krusche** [comment]
> > See if everything works as expected
> 
> What is the expected Score for 0 point exercises? I created a new text exercise with 0 points and 0 bonus points, then assessed a submission with 1 point which results in the following:
> grafik
> I assume that is correct because we use a fallback of `100` as base when there are no points. However, on the course statistics for the student I still have 0 points? (When in fact I should have 1 point?)
> grafik

In this case it should always state 0%. You cannot achieve any points, so a score different from 0% would not make sense at the moment. (We can think about supporting this differently in the future, but such exercises do not occur often)

**Review by kloessst — CHANGES_REQUESTED** [review]
Everything worked fine. I added small functional suggestions in a comment above. A few notes about the code:

**Comment by stefanwaldhauser** [comment]
> We can think about supporting this differently in the future, but such exercises do not occur often)


> Seems like complaints are deactivated for exercises without points and bonus. This makes sense but we should explain this in the tooltip:

This is not related to the points it all. The code is in complaint-interactions.component.ts Please check again that this is really a problem only for  0 0 exercises..
image

**Comment by stefanwaldhauser** [comment]
> > See if everything works as expected
> 
> What is the expected Score for 0 point exercises? I created a new text exercise with 0 points and 0 bonus points, then assessed a submission with 1 point which results in the following:
> grafik
> I assume that is correct because we use a fallback of `100` as base when there are no points. However, on the course statistics for the student I still have 0 points? (When in fact I should have 1 point?)
> grafik



> > > See if everything works as expected
> > 
> > 
> > What is the expected Score for 0 point exercises? I created a new text exercise with 0 points and 0 bonus points, then assessed a submission with 1 point which results in the following:
> > grafik
> > I assume that is correct because we use a fallback of `100` as base when there are no points. However, on the course statistics for the student I still have 0 points? (When in fact I should have 1 point?)
> > grafik
> 
> In this case it should always state 0%. You cannot achieve any points, so a score different from 0% would not make sense at the moment. (We can think about supporting this differently in the future, but such exercises do not occur often)

It is unclear to me how this is an error. As far as I understood the PR the whole point of 100 point workaround was that students still see a score even though they receive no points?

**Comment by krusche** [comment]
After communication issues and after finding out that the changes in this PR include some wrong implementation, we decided to close this PR and further work on an improved solution in a new PR https://github.com/ls1intum/Artemis/pull/2651

**Pull request closed without merging**